File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/palindrome-products/palindrome-products.js
//
// This is only a SKELETON file for the 'Palindrome Products' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Palindromes {
  static generate({ minFactor, maxFactor }) {
    if (minFactor === undefined || maxFactor === undefined) {
      throw new Error('minFactor and maxFactor must be specified');
    }
    if (minFactor > maxFactor) {
      throw new Error('min must be <= max');
    }

    const isPalindrome = (n) => {
      const s = String(n);
      return s === s.split('').reverse().join('');
    };

    const palindromes = {};

    for (let a = minFactor; a <= maxFactor; a++) {
      for (let b = a; b <= maxFactor; b++) {
        const product = a * b;
        if (isPalindrome(product)) {
          if (!palindromes[product]) {
            palindromes[product] = [];
          }
          palindromes[product].push([a, b]);
        }
      }
    }

    const palindromeValues = Object.keys(palindromes).map(Number);
    if (palindromeValues.length === 0) {
      return {
        smallest: { value: null, factors: [] },
        largest: { value: null, factors: [] }
      };
    }

    const smallestValue = Math.min(...palindromeValues);
    const largestValue = Math.max(...palindromeValues);

    return {
      smallest: { value: smallestValue, factors: palindromes[smallestValue] },
      largest: { value: largestValue, factors: palindromes[largestValue] }
    };
  }
}

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/palindrome-products/palindrome_products.go
package palindrome

import (
	"errors"
	"strconv"
)

type Product struct {
	Value          int
	Factorizations [][2]int
}

func isPalindrome(n int) bool {
	s := strconv.Itoa(n)
	for i := 0; i < len(s)/2; i++ {
		if s[i] != s[len(s)-1-i] {
			return false
		}
	}
	return true
}

func Products(fmin, fmax int) (Product, Product, error) {
	if fmin > fmax {
		return Product{}, Product{}, errors.New("fmin > fmax")
	}

	products := make(map[int][][2]int)

	for a := fmin; a <= fmax; a++ {
		for b := a; b <= fmax; b++ {
			p := a * b
			if isPalindrome(p) {
				products[p] = append(products[p], [2]int{a, b})
			}
		}
	}

	if len(products) == 0 {
		return Product{}, Product{}, errors.New("no palindromes")
	}

	var minP, maxP int
	first := true
	for p := range products {
		if first {
			minP = p
			maxP = p
			first = false
			continue
		}
		if p < minP {
			minP = p
		}
		if p > maxP {
			maxP = p
		}
	}

	smallest := Product{Value: minP, Factorizations: products[minP]}
	largest := Product{Value: maxP, Factorizations: products[maxP]}

	return smallest, largest, nil
}
